cpp/src/arrow/compute/kernels/aggregate_basic_internal.h

```diff
@@ -549,6 +549,28 @@ struct BooleanMinMaxImpl : public MinMaxImpl<BooleanType, SimdLevel> {
   }
 };
 
+template <SimdLevel::type SimdLevel>
+struct BooleanAnyImpl : public MinMaxImpl<BooleanType, SimdLevel> {
+  using MinMaxImpl<BooleanType, SimdLevel>::MinMaxImpl;
+
+  void Consume(KernelContext*, const ExecBatch& batch) override {
+    // short-circuit if seen a True already
+    if (this->state.max == true) {
+      return;
+    }
+
+    BooleanArray arr(batch[0].array());
+    const auto true_count = arr.true_count();
```

Comment:
This forces counting the set bits in the whole array, but we're only interested in the presence of a single set bit, so we should be able to shortcut much more aggressively.

You may try to use `OptionalBinaryBitBlockCounter` for that. Untested:
```c++
const auto& data = *batch[0].array();
OptionalBinaryBitBlockCounter counter(data.buffers[0], data.offset, data.buffers[1], data.offset, data.length);
int64_t position = 0;
while (position < data.length) {
  const auto block = counter.NextAndBlock();
  if (block.popcount > 0) {
    this->state.max = true;
    break;
  }
  position += block.length;
}
```
}